src/Hamiltonians/HubbardRealSpace.jl

```diff
@@ -276,116 +308,160 @@ starting_address(h::HubbardRealSpace) = h.address
 
-# Multi-component part
-"""
-    HubbardRealSpaceOffdiagonals{A,T<:Tuple} <: AbstractOffdiagonals{A,Float64}
+function diagonal_element(col::HubbardRealSpaceColumn)
+    h = col.hamiltonian
+    occmaps = map(c -> c.occmap, col.components)
+    int = isnothing(h.u) ? 0.0 : local_interaction(col.address, h.u, occmaps)
+    pot = isnothing(h.v) ? 0.0 : external_potential(col.address, h.potential, occmaps)
 
-Offdiagonals of a [`HubbardRealSpace`](@ref) model with a [`CompositeFS`](@ref) address.
-"""
-struct HubbardRealSpaceOffdiagonals{A,T<:Tuple} <: AbstractOffdiagonals{A,Float64}
-    address::A
-    parts::T
-    length::Int
+    return int + pot
 end
 
-"""
-    get_comp_offdiags(h::HubbardRealSpace, add)
+function operator_column(h::HubbardRealSpace, address)
+    components = _column_components(h, address)
+    return HubbardRealSpaceColumn(
+        h, h.geometry, address, components, sum(length, components)
+    )
+end
 
-Get offdiagonals of all components of address in a type-stable manner.
-"""
-@inline function get_comp_offdiags(h::HubbardRealSpace, address)
-    return _get_comp_offdiags(address.components, h, Val(1))
+# Collect HubbardRealSpaceComponentData for each component of the address.
+@inline function _column_components(h::HubbardRealSpace, address::SingleComponentFockAddress)
+    return (HubbardRealSpaceComponentData{1}(h.geometry, address, address, h.t[1]),)
+end
+@inline function _column_components(h::HubbardRealSpace, address::CompositeFS)
+    return _column_components(h, address, address.components, Val(1))
+end
+@inline function _column_components(::HubbardRealSpace, _, ::Tuple{}, ::Val)
+    return ()
+end
+@inline function _column_components(
+    h::HubbardRealSpace, address, (a, as...), ::Val{I}
+) where {I}
+    data = HubbardRealSpaceComponentData{I}(h.geometry, address, a, h.t[I])
+    rest = _column_components(h, address, as, Val(I + 1))
+    return (data, rest...)
 end
 
-# All steps of recursive function (should) get inlined, creating a type-stable tuple of
-# offdiagonals.
-@inline function _get_comp_offdiags((a,as...), h, ::Val{I}) where {I}
-    return (offdiagonals(h, I, a), _get_comp_offdiags(as, h, Val(I+1))...)
+function _split_index_component(column, i)
+    components = column.components
+    chosen_component = 0
+    while i > 0
+        chosen_component += 1
+        i -= index_apply(length, components, chosen_component)
+    end
+    i += index_apply(length, components, chosen_component)
+    return chosen_component, i
 end
-@inline _get_comp_offdiags(::Tuple{}, h, ::Val) = ()
 
-function offdiagonals(h::HubbardRealSpace{C,A}, address::A) where {C,A<:CompositeFS}
-    parts = get_comp_offdiags(h, address)
-    return HubbardRealSpaceOffdiagonals(address, parts, sum(length, parts))
+function random_offdiagonal(column::HubbardRealSpaceColumn)
+    directions = 2 * num_dimensions(column.hamiltonian.geometry)
+    random_number = rand(1:num_offdiagonals(column))
```

Comment:
```suggestion
    random_number = rand(1:column.num_offdiagonals)
```
for consistency with line 424? It makes it clearer that it's a precomputed value.